API/Data/Repository/MessageRepository.cs

```diff
@@ -0,0 +1,112 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using API.Data;
+using AutoMapper;
+using AutoMapper.QueryableExtensions;
+using Data.Repository.Interface;
+using DTOs;
+using Entities;
+using Helpers;
+using Microsoft.EntityFrameworkCore;
+
+namespace Data.Repository
+{
+    public class MessageRepository : IMessageRepository
+    {
+        private readonly DataContext _context;
+        private readonly IMapper _mapper;
+        
+        public MessageRepository(
+            DataContext context,
+            IMapper mapper
+        ) {
+            _context = context;
+            _mapper = mapper;
+        }
+        public void AddMessage(Message message)
+        {
+            _context.Messages.Add(message);
+        }
+
+        public void DeleteMessage(Message message)
+        {
+            _context.Messages.Remove(message);
+        }
+
+        public async Task<Message> GetMessage(int id)
+        {
+            return await _context.Messages
+                .Include(u => u.Sender)
+                .Include(u => u.Recipient)
+                .SingleOrDefaultAsync(x => x.Id == id);
+        }
+
+        public async Task<PagedList<MessageDTO>> GetMessageForUser(MessageParams messageParams)
+        {
+            var query = _context.Messages
+                .OrderByDescending(m => m.MessageSent)
+                .AsQueryable();
+
+            query = messageParams.Container switch
+            {
+                "Inbox" => query.Where(u => u.Recipient.UserName == messageParams.Username
+                    && u.RecipientDelted == false),
+                "Outbox" => query.Where(u => u.Sender.UserName == messageParams.Username
+                    && u.SenderDelted == false),
+                _ => query.Where(u => u.Recipient.UserName == messageParams.Username
+                    && u.RecipientDelted == false
+                    && u.DateRead == null)
+            };
+
+            var messages = query.ProjectTo<MessageDTO>(_mapper.ConfigurationProvider);
+
+            return await PagedList<MessageDTO>.CreateAsync(
+                messages, 
+                messageParams.PageNumber, 
+                messageParams.PageSize);
+        }
+
+        public async Task<IEnumerable<MessageDTO>> GetMessageThread(
+            string currentUsername,
+            string recipientUsername
+        ) {
+            var messages = await _context.Messages
```

Comment:
For a query that requires dependent entity's info, suggest to create a custom model rather than including all the fields. The reason is query for more fields, more impact on the database performance.
With a custom model, you just need to add one line before the ToListAsync(): .Select(m => new [Your Model] )